lib/logstash/outputs/elasticsearch/http_client/pool.rb

```diff
@@ -0,0 +1,337 @@
+
+      logger.debug("Closing adapter #{@adapter}")
+      @adapter.close
+    end
+
+    def wait_for_in_use_connections
+      until in_use_connections.empty?
+        logger.info "Blocked on shutdown to in use connections #{@state_mutex.synchronize {@url_info}}"
+        sleep 1
+      end
+    end
+
+    def in_use_connections
+      @state_mutex.synchronize { @url_info.values.select {|v| v[:in_use] > 0 } }
+    end
+
+    def alive_urls_count
+      @state_mutex.synchronize { @url_info.values.select {|v| !v[:dead] }.count }
+    end
+
+    def until_stopped(task_name)
+      until @state_mutex.synchronize { @stopping }
+        begin
+          yield
+        rescue => e
+          logger.warn(
+            "Error while performing #{task_name}",
+            :message => e.message,
+            :class => e.class.name,
+            :backtrace => e.backtrace
+            )
+        end
+      end
+    end
+
+    def start_sniffer
+      @sniffer = Thread.new do
+        until_stopped("sniffing") { sniff! }
+      end
+    end
+
+    ES1_SNIFF_RE_URL  = /\[([^\/]*)?\/?([^:]*):([0-9]+)\]/
+    ES2_SNIFF_RE_URL  = /([^\/]*)?\/?([^:]*):([0-9]+)/
+    def sniff!
+      resp = perform_request('GET', '_nodes')
+      parsed = LogStash::Json.load(resp.body)
+      parsed['nodes'].map do |id,info|
+        # TODO Make sure this works with shield. Does that listed
+        # stuff as 'https_address?'
+        addr_str = info['http_address'].to_s
+        next unless addr_str # Skip hosts with HTTP disabled
+
+
+        # Only connect to nodes that serve data
+        # this will skip connecting to client, tribe, and master only nodes
+        # Note that if 'attributes' is NOT set, then that's just a regular node
+        # with master + data + client enabled, so we allow that
+        attributes = info['attributes']
+        next if attributes && attributes['data'] == 'false'
+
+        matches = addr_str.match(ES1_RE_URL) || addr_str.match(ES2_RE_URL)
+        if matches
+          host = matches[1].empty? ? matches[2] : matches[1]
+          port = matches[3]
+          info.merge :host => host, :port => port, :id => id
+        end
+      end.compact
+    end
+
+    def stop_sniffer
+      @sniffer.join if @sniffer
+    end
+
+    def start_resurrectionist
+      @resurrectionist = Thread.new do
+        last_resurrect = Time.now
+        until_stopped("resurrecting") do
+          if Time.now-last_resurrect >= @resurrect_interval
+            last_resurrect = Time.now
+            resurrect_dead!
+          end
+        end
+      end
+    end
+
+    def resurrect_dead!
+      # Try to keep locking granularity low such that we don't affect IO...
+      @state_mutex.synchronize { @url_info.select {|url,meta| meta[:dead] } }.each do |url,meta|
+        begin
+          @logger.info("Checking url #{url} with path #{@healthcheck_path} to see if node resurrected")
+          perform_request_to_url(url, "GET", @healthcheck_path)
```

Comment:
since we're not performing any validation on the response, I wonder if we could switch to HEAD.